src/calc_integrate.jl

```diff
@@ -0,0 +1,325 @@
+        @libarb(acb_calc_integrate),
+        Cint,
+        (
+            Ref{acb_struct},
+            Ptr{Cvoid},
+            Any,
+            Ref{acb_struct},
+            Ref{acb_struct},
+            Clong,
+            Ref{mag_struct},
+            Ptr{Cvoid},
+            Clong,
+        ),
+        res,
+        acb_calc_func,
+        param,
+        a,
+        b,
+        rel_goal,
+        abs_tol,
+        options,
+        prec,
+    )
+end
+
+"""
+    _integrate!(res, f, a, b; prec, rel_goal, abs_goal, opts)
+
+Internal function for integration. Assumes that `f!` is of the type
+`f!(out::AcbRef, inp::AcbRef; analytic::Bool, prec::Integer)` and
+that all other parameters are set to valid values.
+"""
+function _integrate!(
+    f!,
+    res::AcbLike,
+    a::AcbLike,
+    b::AcbLike;
+    prec::Integer = precision(res),
+    rel_goal::Integer = prec,
+    abs_tol::MagLike = set_ui_2exp!(Arblib.Mag(), one(UInt), -prec),
+    opts::Union{Ptr{Cvoid},calc_integrate_opt_struct} = C_NULL,
+)
+    status = calc_integrate!(
+        res,
+        _acb_calc_cfunc!(),
+        f!,
+        a,
+        b,
+        rel_goal,
+        abs_tol,
+        opts, # passing C_NULL uses the default options
+        prec,
+    )
+
+    # status:
+    # ARB_CALC_SUCCESS = 0
+    # ARB_CALC_NO_CONVERGENCE = 2
+    status == 2 && @warn "integration did not converge"
+
+    return res
+end
+
+"""
+    integrate!(f!, res::Acb, a::Number, b::Number;
+        check_analytic::Bool = false,
+        take_prec::Bool = false,
+        prec::Integer = precision(res),
+        rtol = 0.0,
+        atol = 2.0^-prec,
+        opts::Union{acb_calc_integrate_opt_struct, Ptr{Cvoid}} = C_NULL,
+    )
+
+Like [`integrate`](@ref), but make use of in-place operations. In
+particular, there are three differences from `integrate`:
+
+1. The function `f!` should be of the form `f!(y, x) = set!(y, f(x))`.
+   That is, it writes the return value of the integand `f(x)` in-place
+   into its first argument `y`. (The return value of `f!` is ignored).
+
+2. `res` is set to the result.
+
+3. The default precision is taken from `res` instead of from `a` and
+   `b`.
+"""
+function integrate!(
+    f!,
+    res::Union{Acb,AcbRef},
+    a::Union{Acb,AcbRef},
+    b::Union{Acb,AcbRef};
+    check_analytic::Bool = false,
+    take_prec::Bool = false,
+    prec::Integer = precision(res),
+    rtol = 0.0,
+    atol = set_ui_2exp!(Mag(), one(UInt), -prec),
+    opts::Union{Ptr{Cvoid},calc_integrate_opt_struct} = C_NULL,
+)
+    # rtol ≈ 2^(-rel_goal)
+    rel_goal = iszero(rtol) ? prec : max(-floor(Int, log2(atol)), 0)
+
+    if !check_analytic && !take_prec
+        g! = (inp, out; analytic, prec) -> f!(inp, out)
+    elseif !check_analytic && take_prec
+        g! = (inp, out; analytic, prec) -> f!(inp, out, prec = prec)
+    elseif check_analytic && !take_prec
+        g! = (inp, out; analytic, prec) -> f!(inp, out, analytic = analytic)
+    else
+        g! = f!
+    end
```

Comment:
I'm also slightly worried about this construction of `g!` having performance implications. I don't think so, but I'm not sure.